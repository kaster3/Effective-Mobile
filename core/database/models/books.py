from enum import Enum

from sqlalchemy import Enum as SQLEnum
from sqlalchemy import Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from core.database.models.base import Base
from core.database.models.mixins import IntIdPkMixin


class BookStatus(Enum):
    IN_STOCK = "В наличие"
    CHECKED_OUT = "Выдана"


class Book(Base, IntIdPkMixin):
    title: Mapped[str] = mapped_column(String(30), unique=True)
    author: Mapped[str] = mapped_column(String(30))
    year: Mapped[int] = mapped_column(Integer)
    status: Mapped[BookStatus] = mapped_column(SQLEnum(BookStatus))
