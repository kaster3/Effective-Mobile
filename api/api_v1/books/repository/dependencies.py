from typing import Annotated

from fastapi import Depends, HTTPException, Path

from api.api_v1.books.dependencies import get_book_repository
from core.database.models import Book

from .books import BookRepository


async def get_book_by_params(
    book_params: Annotated[str, Path],
    repository: Annotated[BookRepository, Depends(get_book_repository)],
) -> Book:

    if book_params.isdigit():
        book = await repository.get_by_id(book_id=int(book_params))
    else:
        book = await repository.get_by_title(book_title=book_params)

    if book is not None:
        return book
    raise HTTPException(status_code=404, detail="Book not found")
