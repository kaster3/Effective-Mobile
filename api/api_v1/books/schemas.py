from pydantic import BaseModel, ConfigDict

from core.database.models.books import BookStatus


class BaseBook(BaseModel):
    title: str
    author: str
    year: int
    status: BookStatus


class Book(BaseBook):
    model_config = ConfigDict(from_attributes=True)
    id: int


class BookCreate(BaseBook):
    pass


class BookUpdate(BaseBook):
    pass


class BookUpdatePartial(BaseBook):
    title: str | None = None
    author: str | None = None
    year: int | None = None
    status: BookStatus | None = None
