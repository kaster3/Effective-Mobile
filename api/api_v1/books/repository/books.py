import sqlalchemy
from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from api.api_v1.books.schemas import BookCreate, BookUpdatePartial
from core.database.models import Book


class BookRepository:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def get_books(self) -> list[Book]:
        smtp = select(Book).order_by(Book.id)
        books = await self.session.scalars(smtp)
        return list(books)

    async def get_by_id(self, book_id: int) -> Book | None:
        book = await self.session.get(Book, book_id)
        return book

    async def get_by_title(self, book_title: str) -> Book | None:
        smtp = select(Book).where(Book.title == book_title)
        book = await self.session.scalar(smtp)
        return book

    async def create_book(self, book_in: BookCreate) -> Book:
        book = Book(**book_in.model_dump())
        self.session.add(book)
        await self.session.commit()
        await self.session.refresh(book)
        return book

    async def update_book(
        self,
        book: Book,
        book_in: BookUpdatePartial,
    ) -> Book:
        for key, value in book_in.model_dump(exclude_unset=True).items():
            setattr(book, key, value)
        await self.session.commit()
        return book

    async def delete_book(self, book: Book) -> None:
        try:
            await self.session.delete(book)
            await self.session.commit()
        except sqlalchemy.exc.IntegrityError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot delete category"
            )
