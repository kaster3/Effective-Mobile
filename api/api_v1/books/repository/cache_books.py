import json

from redis import asyncio as Redis

from api.api_v1.books.schemas import Book


class CacheRepository:
    def __init__(self, redis: Redis):
        self.redis = redis

    async def get_books(self) -> list[Book]:
        async with self.redis as redis:
            books_json = await redis.lrange("books", 0, -1)

            return [Book.model_validate(json.loads(book)) for book in books_json]

    async def set_books(self, books: list[Book]):
        books_json = [book.json() for book in books]
        async with self.redis as redis:
            await redis.lpush("books", *books_json)
