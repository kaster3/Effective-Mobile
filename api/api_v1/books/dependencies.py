from typing import Annotated

from fastapi import Depends
from redis.asyncio import Redis
from sqlalchemy.ext.asyncio import AsyncSession

from api.api_v1.books.repository.books import BookRepository
from api.api_v1.books.repository.cache_books import CacheRepository
from api.api_v1.books.service import BookService
from core.database.db_helper import db_helper
from core.settings import Settings


async def get_settings() -> Settings:
    return Settings()


async def get_book_repository(
    session: AsyncSession = Depends(db_helper.session_getter),
) -> BookRepository:
    return BookRepository(session=session)


async def get_redis_connection(settings: Settings) -> Redis:
    return await Redis(
        host=settings.redis.host, port=settings.redis.port, db=settings.redis.db
    )


async def get_cache_repository() -> CacheRepository:
    redis_connection = await get_redis_connection(settings=await get_settings())
    return CacheRepository(redis_connection)


async def get_book_service(
    repository: BookRepository = Depends(get_book_repository),
    cache_repository: CacheRepository = Depends(get_cache_repository),
) -> BookService:
    return BookService(repository=repository, cache=cache_repository)
