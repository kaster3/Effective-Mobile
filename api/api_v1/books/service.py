import logging

from fastapi import HTTPException

from api.api_v1.books.repository.books import BookRepository
from api.api_v1.books.repository.cache_books import CacheRepository
from api.api_v1.books.schemas import Book as BookSchema
from api.api_v1.books.schemas import BookCreate, BookUpdatePartial
from core.database.models import Book

log = logging.getLogger(__name__)


class BookService:
    def __init__(self, repository: BookRepository, cache: CacheRepository):
        self.repository = repository
        self.cache = cache

    async def get_books(self) -> list[BookSchema]:
        if cache_books := await self.cache.get_books():
            log.debug("--------------------------------")
            return cache_books
        else:
            books = await self.repository.get_books()
            books_schema = [BookSchema.model_validate(book) for book in books]
            await self.cache.set_books(books_schema)
            return books_schema

    async def create_book(self, book_in: BookCreate) -> Book:
        # Проверяем нет ли уже книги с таким названием так как полу уникально
        if await self.repository.get_by_title(book_in.title):
            raise HTTPException(status_code=400, detail="Book with this title already exists")

        new_book = await self.repository.create_book(book_in=book_in)
        return new_book

    async def update_book(self, book: Book, book_in: BookUpdatePartial) -> Book:
        # Проверяем нет ли уже книги с таким названием так как полу уникально
        if book_in.title and await self.repository.get_by_title(book_in.title):
            raise HTTPException(status_code=400, detail="Book with this title already exists")

        updated_book = await self.repository.update_book(book=book, book_in=book_in)
        return updated_book

    async def delete_book_by_id(self, book_id: int) -> None:
        book = await self.repository.get_by_id(book_id=book_id)
        await self.repository.delete_book(book=book)
