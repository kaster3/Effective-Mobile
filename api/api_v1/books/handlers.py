from typing import Annotated

from fastapi import APIRouter, Depends, status

from core.settings import settings

from .dependencies import get_book_service
from .repository.dependencies import get_book_by_params
from .schemas import Book, BookCreate, BookUpdatePartial
from .service import BookService

router = APIRouter(
    prefix=settings.api.v1.books,
    tags=["books"],
)


@router.get("", response_model=list[Book], status_code=status.HTTP_200_OK)
async def get_books(service: Annotated[BookService, Depends(get_book_service)]):
    books = await service.get_books()
    return books


@router.get("/{book_id_or_title}", response_model=Book, status_code=status.HTTP_200_OK)
async def get_book_by_id(
    book: Annotated[Book, Depends(get_book_by_params)],
):
    return book


@router.post("", response_model=BookCreate, status_code=status.HTTP_201_CREATED)
async def create_book(
    book_in: BookCreate,
    service: Annotated[BookService, Depends(get_book_service)],
):
    new_book = await service.create_book(book_in=book_in)
    return new_book


@router.patch(
    "/{book_id_or_title}", response_model=BookUpdatePartial, status_code=status.HTTP_200_OK
)
async def update_book(
    book: Annotated[Book, Depends(get_book_by_params)],
    book_in: BookUpdatePartial,
    service: Annotated[BookService, Depends(get_book_service)],
):
    updated_book = await service.update_book(book=book, book_in=book_in)
    return updated_book


@router.delete("/{book_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_book_by_id(
    book_id: int,
    service: Annotated[BookService, Depends(get_book_service)],
):
    await service.delete_book_by_id(book_id=book_id)
