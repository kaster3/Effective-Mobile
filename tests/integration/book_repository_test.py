import pytest
from syrupy import SnapshotAssertion, snapshot

from api.api_v1.books.schemas import BookCreate, BookUpdatePartial
from core.database.models import Book
from api.api_v1.books.repository.books import BookRepository
from core.database.models.books import BookStatus


@pytest.mark.asyncio
async def test_get_all_books(init_models, create_books, book_repository: BookRepository, snapshot: SnapshotAssertion):
    books = await book_repository.get_books()
    assert len(books) == 2
    assert isinstance(books[0], Book)
    assert isinstance(books[1], Book)

    snapshot.assert_match([book.to_dict() for book in books])


@pytest.mark.asyncio
async def test_create_book(init_models, book_repository: BookRepository, snapshot: SnapshotAssertion):
    book_1 = await book_repository.create_book(
        BookCreate(title="Test title 1", author="Test author 1", year=1000, status=BookStatus.IN_STOCK)
    )
    book_2 = await book_repository.create_book(
        BookCreate(title="Test title 2", author="Test author 1", year=1000, status=BookStatus.CHECKED_OUT)
    )

    snapshot.assert_match(book_1.to_dict())
    snapshot.assert_match(book_2.to_dict())



@pytest.mark.asyncio
async def test_get_book_by_id(init_models, create_books, book_repository: BookRepository, snapshot: SnapshotAssertion):
    book_1 = await book_repository.get_by_id(1)
    book_2 = await book_repository.get_by_id(2)
    book_3 = await book_repository.get_by_id(3)

    assert book_1 is not None
    assert book_2 is not None
    assert book_3 is None

    snapshot.assert_match(book_1.to_dict())
    snapshot.assert_match(book_2.to_dict())

@pytest.mark.asyncio
async def test_get_book_by_title(init_models, create_books, book_repository: BookRepository, snapshot: SnapshotAssertion):
    book_1 = await book_repository.get_by_title(book_title="Test title 1")
    book_2 = await book_repository.get_by_title(book_title="Test title 2")
    book_3 = await book_repository.get_by_title(book_title="Test title 3")

    assert book_1 is not None
    assert book_2 is not None
    assert book_3 is None

    snapshot.assert_match(book_1.to_dict())
    snapshot.assert_match(book_2.to_dict())


@pytest.mark.asyncio
async def test_update_book(init_models, get_book_by_id: Book, book_repository: BookRepository, snapshot: SnapshotAssertion):
    assert get_book_by_id is not None
    snapshot.assert_match(get_book_by_id.to_dict())

    updated_book = await book_repository.update_book(
        book=get_book_by_id,
        book_in=BookUpdatePartial(
            title="Updated title",
            author="Updated author",
            year=2000,
            status=BookStatus.CHECKED_OUT)
    )

    snapshot.assert_match(updated_book.to_dict())



@pytest.mark.asyncio
async def test_delete_book(init_models, get_book_by_id: Book, book_repository: BookRepository):
    assert get_book_by_id is not None
    assert isinstance(get_book_by_id, Book)
    await book_repository.delete_book(book=get_book_by_id)
    book = await book_repository.get_by_id(get_book_by_id.id)
    assert book is None




