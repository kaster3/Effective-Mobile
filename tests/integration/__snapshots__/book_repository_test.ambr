# serializer version: 1
# name: test_create_book
  dict({
    'author': 'Test author 1',
    'id': 1,
    'status': 'В наличие',
    'title': 'Test title 1',
    'year': 1000,
  })
# ---
# name: test_create_book.1
  dict({
    'author': 'Test author 1',
    'id': 2,
    'status': 'Выдана',
    'title': 'Test title 2',
    'year': 1000,
  })
# ---
# name: test_get_all_books
  list([
    dict({
      'author': 'Test author 1',
      'id': 1,
      'status': 'В наличие',
      'title': 'Test title 1',
      'year': 1000,
    }),
    dict({
      'author': 'Test author 2',
      'id': 2,
      'status': 'Выдана',
      'title': 'Test title 2',
      'year': 1000,
    }),
  ])
# ---
# name: test_get_book_by_id
  dict({
    'author': 'Test author 1',
    'id': 1,
    'status': 'В наличие',
    'title': 'Test title 1',
    'year': 1000,
  })
# ---
# name: test_get_book_by_id.1
  dict({
    'author': 'Test author 2',
    'id': 2,
    'status': 'Выдана',
    'title': 'Test title 2',
    'year': 1000,
  })
# ---
# name: test_get_book_by_title
  dict({
    'author': 'Test author 1',
    'id': 1,
    'status': 'В наличие',
    'title': 'Test title 1',
    'year': 1000,
  })
# ---
# name: test_get_book_by_title.1
  dict({
    'author': 'Test author 2',
    'id': 2,
    'status': 'Выдана',
    'title': 'Test title 2',
    'year': 1000,
  })
# ---
# name: test_update_book
  dict({
    'author': 'Test author 1',
    'id': 1,
    'status': 'В наличие',
    'title': 'Test title 1',
    'year': 1000,
  })
# ---
# name: test_update_book.1
  dict({
    'author': 'Updated author',
    'id': 1,
    'status': 'Выдана',
    'title': 'Updated title',
    'year': 2000,
  })
# ---
